pep-0xxx.rst

```diff
@@ -0,0 +1,1250 @@
+
+            if highest_supported_version is _DEFAULT_VALUE:
+                highest_supported_version = self.highest_supported_version
+
+            if trust_store is _DEFAULT_VALUE:
+                trust_store = self.trust_store
+
+            if sni_callback is _DEFAULT_VALUE:
+                sni_callback = self.sni_callback
+
+            return self.__class__(
+                validate_certificates, certificate_chain, ciphers,
+                inner_protocols, lowest_supported_version,
+                highest_supported_version, trust_store, sni_callback
+            )
+
+
+
+Context
+~~~~~~~
+
+We define two Context abstract base classes. These ABCs define objects that
+allow configuration of TLS to be applied to specific connections. They can be
+thought of as factories for ``TLSWrappedSocket`` and ``TLSWrappedBuffer``
+objects.
+
+Unlike the current ``ssl`` module, we provide two context classes instead of
+one. Specifically, we provide the ``ClientContext`` and ``ServerContext``
+classes. This simplifies the APIs (for example, there is no sense in the server
+providing the ``server_hostname`` parameter to ``ssl.SSLContext.wrap_socket``,
+but because there is only one context class that parameter is still available),
+and ensures that implementations know as early as possible which side of a TLS
+connection they will serve. Additionally, it allows implementations to opt-out
+of one or either side of the connection. For example, SecureTransport on macOS
+is not really intended for server use and has an enormous amount of
+functionality missing for server-side use. This would allow SecureTransport
+implementations to simply not define a concrete subclass of ``ServerContext``
+to signal their lack of support.
+
+As much as possible implementers should aim to make these classes immutable:
+that is, they should prefer not to allow users to mutate their internal state
+directly, instead preferring to create new contexts from new TLSConfiguration
+objects. Obviously, the ABCs cannot enforce this constraint, and so they do not
+attempt to.
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+
+
+    class _BaseContext(metaclass=ABCMeta):
+        @abstractmethod
+        def __init__(self, configuration: TLSConfiguration):
+            """
+            Create a new context object from a given TLS configuration.
+            """
+
+        @property
+        @abstractmethod
+        def configuration(self) -> TLSConfiguration:
+            """
+            Returns the TLS configuration that was used to create the context.
+            """
+
+
+    class ClientContext(_BaseContext):
+        @abstractmethod
+        def wrap_socket(self,
+                        socket: socket.socket,
+                        server_hostname: Optional[str],
+                        auto_handshake=True: bool) -> TLSWrappedSocket:
```

Comment:
Yeah, it's definitely worth considering. It makes this a bit more of a diversion from the most-common approach of the `ssl` module, but that's probably ok.